### PR TITLE
NDRS-1082: Modify hashing of the block body

### PR DESCRIPTION
**Pull request by fizyk20**
This changes the hashing scheme for block bodies in a way supporting Merkle proofs.

https://casperlabs.atlassian.net/browse/NDRS-1082

### CODE REVIEW

File: node/src/types/block.rs
Comment (goral09) [inline node/src/types/block.rs:899]: :heart_eyes: :tada: :bow: :1st_place_medal:

File: node/src/types/block.rs
Comment (goral09) [inline node/src/types/block.rs:931]: To me it looks like a fold left over the collection of hashes, matter of taste of course.

File: node/src/types/block.rs
Comment (goral09) [inline node/src/types/block.rs:924]: Maybe it's a weird question but why are you filling the array "from the end"?

Comment (fizyk20) [reply node/src/types/block.rs:924]: In order to preserve the structure of the tree. `rfold` takes elements one by one from the right, and the result of the previous call is the _first_ argument, not the second, so the order appears to be reversed.

For example, the first call will get `(hashed_empty, hashed_proposer)`, and I'm supposed to hash `hashed_proposer + hashed_empty`.

Comment (goral09) [reply node/src/types/block.rs:924]: I mean, if those two lines were simply reordered? Without changing what gets written where.

Comment (fizyk20) [reply node/src/types/block.rs:924]: Ah, but if I understand you correctly, the first line actually fills the first part of the array, and the second line fills the second part. So it's not actually filling "from the end", and that's why it's in this order.

Or do you just prefer the line using `prev` to be the first one?

**Review by afck — COMMENTED** [review]
The hashing looks correct to me, but can we make that change just like that? Will it work with upgrades? Specifically: What if a node tries to join the upgraded network; will it be able to download the pre-upgrade blocks? I guess it will, because they are stored by their old hash, and we never recompute their hash after the upgrade?

File: node/src/types/block.rs
Comment (goral09) [reply node/src/types/block.rs:924]: :see_no_evil: :bat: My bad. Sorry. It's OK, just like I like it :)

**Comment by fizyk20** [comment]
Right, I believe it should work with upgrades... as long as we don't try to use any logic referring to the Merkle proofs on older blocks. Although, I wonder if we can be sure that we won't do that :grimacing:

**Comment by goral09** [comment]
> Right, I believe it should work with upgrades... as long as we don't try to use any logic referring to the Merkle proofs on older blocks. Although, I wonder if we can be sure that we won't do that grimacing

Easy, just run the network with two binaries running alongside.

**Review by xcthulhu — APPROVED** [review]

File: node/src/crypto/hash.rs
Comment (goral09) [inline node/src/crypto/hash.rs:254]: We should add unit tests for these new hashing functions so that no one ever changes the order in which elements are being hashed (for example in `hash_pair` function put `hash1` as the second element of the array, instead of the first or in `hash_btree_map` push elements at the front rather than back).

File: node/src/components/consensus/consensus_protocol.rs
Comment (afck) [inline node/src/components/consensus/consensus_protocol.rs:72]: Maybe simpler:
```suggestion
            let hashes = slice_of_validators.iter().map(|validator| {
                    validator.to_bytes().expect("Could not serialize validator")
            }).collect();
            hash::hash_vec_merkle_tree(hashes)
```

File: node/src/crypto/hash.rs
Comment (afck) [inline node/src/crypto/hash.rs:184]:
```suggestion
/// Hashes a [Vec<Digest>] into a single [Digest] by constructing a [Merkle tree][1].
```
(Also below.)

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:761]: Do we need to keep the old hash function for compatibility, or will it never be used after the upgrade?
E.g. we currently still read [the last block before the upgrade](https://github.com/CasperLabs/casper-node/blob/34c802333b88880ed7fd31f412d222dfda4519ac/node/src/effect.rs#L1568), not to take the validators directly from it, but to find out the height of the first block after the upgrade.

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:777]: (We had talked about maybe using all zeros instead of `hash(&[])`. In this case it doesn't matter, though, because `era_end.hash()` probably can't be `hash(&[])`.)

Comment (goral09) [reply node/src/types/block.rs:777]: FWIW, in other places where we have this kind of "zero element hash" we hash all zeros array.

Maybe, we can have a separate module for these new functions and add a constant `const EMPTY_HASH = hash::hash([0u8; 32])` so that people are not tempted to use them in different contexts? This pattern seems to be used throughout the EE codebase as well where the functions are grouped together and there are constants that hide internal details/design decisions.

File: node/src/types/block.rs
Comment (fizyk20) [reply node/src/types/block.rs:761]: That's a good question. In the CEP, I indicated that we should probably keep the old hash function, so that at least we can validate old storage. I'm not sure how critical that is, though.

File: node/src/crypto/hash.rs
Comment (goral09) [inline node/src/crypto/hash.rs:224]: I just found today that there's a similar method in `itertools` lib for this: https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.tree_fold1

Comment (afck) [reply node/src/crypto/hash.rs:224]: Awesome, I didn't know that!
If we use that we should just have a unit test that makes sure the exact behavior of `tree_fold1` doesn't change, i.e. the shape of a tree with ≠ 2<sup>n</sup> leaves.

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:773]: Isn't that equivalent to:
```suggestion
            .expect("should serialize block header");
```
Or does that use `Debug` instead of `Display`? (Also below.)

**Review by afck — COMMENTED** [review]
There still seem to be build errors.
But since this targets 1.3.0, maybe let's merge it only after the 1.2.0 branch has been cut anyway?

File: node/src/types/block.rs
Comment (fizyk20) [reply node/src/types/block.rs:773]: I've checked, that uses `Debug`.

File: node/src/crypto/hash.rs
Comment (fizyk20) [reply node/src/crypto/hash.rs]: I've made it use `itertools`, and there are tests checking hashing of vectors with odd and even number of elements, so if anything changes, they should start failing.

**Comment by fizyk20** [comment]
> But since this targets 1.3.0, maybe let's merge it only after the 1.2.0 branch has been cut anyway?
I see no problems with that (other than: my ticket will roll :grimacing: ).

File: node/src/components/storage.rs
Comment (afck) [inline node/src/components/storage.rs:330]: Should we have a similar test with a protocol version belonging to the v2 hashes?

**Review by afck — APPROVED** [review]
The code looks good to me. But let's make sure it works fine with upgrades before merging it.

**Review by EdHastingsCasperLabs — CHANGES_REQUESTED** [review]
Blocking until it is 100% confirmed that this is upgrade compatible.

@Fraser999 @sacherjj can one of you instruct @fizyk20 on steps he can take to verify if this logic will work across upgrades? Thanks!

File: node/src/crypto/hash.rs
Comment (goral09) [inline node/src/crypto/hash.rs:261]: I'm wondering whether we shouldn't change the right fold to the left fold here (and in the vec/tree as well). The right fold has a property that new elements are appended (vs prepending in the left fold) which could make it more performant if we ever need to add new elements into it. I'm thinking about a vector of elements (digests) that we extend – prepending would most probably require allocating and rewriting all content of a vector into new one, while appending would only extend the size of the vector without rewriting the content.

Comment (afck) [reply node/src/crypto/hash.rs:261]: I think the idea was that once you know that the era report is the _second_ field, you'll always be able to check proofs of the era report based on the hash, even if other fields get _appended_ in a later protocol version. It's true that appending changes all those hashes, but the length of the era report proof will always be 2. For that you need an `rfold`.

File: node/src/components/storage.rs
Comment (fizyk20) [reply node/src/components/storage.rs:330]: I've actually had to change this, or it didn't work :+1:

**Comment by fizyk20** [comment]
I tested this both manually and using the `sync_upgrade_test.sh` now. `sync_upgrade_test.sh` can be used for this, since it runs an upgrade from 1.0.0 to 2.0.0: 1.0.0 uses the old way of hashing, 2.0.0 the new way (since it's greater than 1.3.0). Because of that I think a new test is not even needed.
The tests actually uncovered a bug, which has been fixed in the last commit.
Is it enough to unblock the PR @EdHastingsCasperLabs ?

**Comment by goral09** [comment]
> I tested this both manually and using the `sync_upgrade_test.sh` now. `sync_upgrade_test.sh` can be used for this, since it runs an upgrade from 1.0.0 to 2.0.0: 1.0.0 uses the old way of hashing, 2.0.0 the new way (since it's greater than 1.3.0). Because of that I think a new test is not even needed.
> The tests actually uncovered a bug, which has been fixed in the last commit.
> Is it enough to unblock the PR @EdHastingsCasperLabs ?

Can you add a comment to that integration test that people should not change the protocol version in the future b/c it has to be greater than X.Y.Z ?

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Comment by fizyk20** [comment]
bors merge

**Pull request merged**
